### DISCUSSION TITLE
Parallel GPU visualization example of trame ?

### DISCUSSION THREAD

**Discussion by nyue** [Kitware/trame#627] [Q&A]
Hi,

  I have been given a brief (1 week) access to a cluster of GPUs with a separate head/jumpbox node

  I am wondering if there are existing example I can use to showcase Trame in such a cluster, hopefully using some volume rendering example.

  I am happy to roll up my sleeve to modify/change example to adapt but I am hoping to start with a known working example.

Kind regards

**Comment by jourdain** [comment]
For parallel rendering, you would need to use ParaView 5.13. If you already have a dataset, load it in ParaView, tweak the rendering to your liking, and save the state as a Python file. 

Then add the following lines into a `trame-app.py` file:

```python
import paraview.web.venv
import ptc

# [... python state file content...]

web_app = ptc.Viewer(from_state=True)
web_app.start()
```

Then on your Cluster, create a venv using the same Python as your ParaView (3.10)

```
python3.10 -m venv /test/pv-venv
source /test/pv-venv
pip install paraview-trame-components
```

Then start your MPI job running pvbatch that 

```
export PV_VENV=/test/pv-venv
/paraview/bin/mpi [...] /paraview/bin/pvbatch /path/to/your/trame-app.py
```

HTH

**Comment by nyue** [comment]
Thanks for the suggestion. Learning curve too steep to wrangle this into a working example as my time is limited. Ended up choosing to go via the traditional mpiexec+pvserver on multiple hosts to showcase.

If I get an extension, I may revisit.

**Reply by jourdain** [reply]
You could call `paraview.simple.Connect(...)` in a trame script to connect to your `mpiexec+pvserver`.

**Reply by nyue** [reply]
Ah! That is interesting, I will look into that.

**Comment by nyue** [comment]
To keep things simple, I am testing locally to get a hang of trame

On my local box, I have /home/nyue/systems/ParaView/5.13.1/ParaView-5.13.1-egl-MPI-Linux-Python3.10-x86_64/bin

I start a session with the following
```
 ./mpiexec -hosts head0 ./pvserver
```

I created the following virtual env via conda
```
name: trame
channels:
  - conda-forge
  - default
dependencies:
  - python=3.10.13
  - paraview=5.13.1
  - trame
  - trame-vtk
  - trame-vuetify
```

I tried a simple connection with the following script

```py
from paraview.simple import *

# Connect to the pvserver
server = servermanager.Connect("head0", 11111)

# Load your data - replace `GetSources()` with appropriate pipeline creation
data_source = GetSources()  # Or replace with specific data load logic

# Example of loading a built-in example if no data source is present
if not data_source:
    data_source = Wavelet()  # Example built-in data source for testing
```

I connect via the created conda environment I named `trame`
```
(trame) nyue@head0:~/temp/trame/simple$ python mypvserver.py 
Segmentation fault (core dumped)
```

I am getting the following error
```
Waiting for client...
Connection URL: cs://head0:11111
Accepting connection(s): head0:11111
Client connected.
(   6.597s) [pvserver        ]vtkSocketCommunicator.c:774    ERR| vtkSocketCommunicator (0x9c45c0): Could not receive tag. 1
(   6.598s) [pvserver        ]vtkTCPNetworkAccessMana:284    ERR| vtkTCPNetworkAccessManager (0x8d3790): Some error in socket processing.
Exiting...
```

**Comment by jourdain** [comment]
You can not use the paraview from conda as it is incompatible with your pvserver. You need to use `pvpython` from the ParaView 5.13.1 that you download from our website. 

Then create a venv from a python 3.10 with all the trame dependencies and run it like below

`pvpython --venv /path/to/venv /path/to/trame-app.py`

**Reply by nyue** [reply]
Understood.

I have since made some head way, I tested it on my laptop and now trying to deploy to the server

```
(.venv) ubuntu@san8oaict:~/projects$ pvbatch state-loader.py --server
ParaView is using venv: /home/ubuntu/projects/.venv

App running at:
 - Local:   http://localhost:8080/
 - Network: http://127.0.0.1:8080/
```

I was wondering if there is a way for the server to listen on host/IP 0.0.0.0 ? like it does with http.server

```
(.venv) ubuntu@san8oaict:~/projects$ python3 -m http.server
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
38.88.120.18 - - [02/Nov/2024 19:26:37] "GET / HTTP/1.0" 200 -
^C
Keyboard interrupt received, exiting.
```

Thank you.

**Reply by nyue** [reply]
Answering my own question, I can use --host 0.0.0.0 to achieve that
